src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs

```diff
@@ -56,41 +72,81 @@ public IEnumerable<DataListItem> GetItems()
             if (json == null)
                 return items;
 
-            foreach (JProperty token in json.Children())
+            try
             {
-                items.Add(new DataListItem
+                var jsonItems = json.SelectTokens(ItemsJsonPath);
+
+                // if either all names are null or values are null, then we'll return empty array and log an exception
+                var allNameNull = true;
+                var allValueNull = true;
+
+                foreach (var item in jsonItems)
                 {
-                    Name = token.Value?.ToString(),
-                    Value = token.Name
-                });
+                    var name = item.SelectToken(NameJsonPath);
+                    var value = item.SelectToken(ValueJsonPath);
+                    var icon = string.IsNullOrEmpty(IconJsonPath) ? null : item.SelectToken(IconJsonPath);
+                    var description = string.IsNullOrEmpty(DescriptionJsonPath) ? null : item.SelectToken(DescriptionJsonPath);
+
+                    // How should we log if either name or value is empty? Note that empty or missing values are totally legal according to json
+                    if (name == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'name' was found using JSONPath: " + NameJsonPath);
+                    if (value == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'value' was found using JSONPath: " + ValueJsonPath);
+
+                    allNameNull = allNameNull && (name == null); // Will flip first time name is not null
+                    allValueNull = allValueNull && (value == null); // Will flip first time value is not null
```

Comment:
Can you clarify your thinking behind tracking if all the names/values are `null`?

With the XML data-source, we check that both the `name` and `value` aren't `null` before adding it to the list.

Reply:
> Can you clarify your thinking behind tracking if all the names/values are `null`?
> 
> With the XML data-source, we check that both the `name` and `value` aren't `null` before adding it to the list.

Yes. I realised during my tests that I had some weird experiences with the dropdown editor. It default-selected the last option, and I couldn't grasp why. Turned out that all my value-fields were blank due to an incorrect jsonpath, which caused this error. Basically it should be ok that some of the name or value fields are empty, technically speaking this is the idea of json; that it's not too strict, but if all names or values are empty, then I would consider that an error and log accordingly.

Makes sense?

Reply:
OK, that's fine. 👍